### ISSUE TITLE
Designing a GUI using Gradio

### ISSUE THREAD

**Issue by vandrw** [lucidrains/alphafold3-pytorch#221]
I have briefly used Gradio before and could have an attempt at designing a frontend for the project. However, before starting, I wanted to discuss how much feature parity we could achieve with DeepMind's AlphaFold server.

Currently, their implementation supports taking as input sequences/fastas of proteins, DNA, RNA, ligands and ions. Should I keep anything in mind while trying to replicate the following in Gradio? I will definitely be coming back with more questions regarding the output of the model, but for now I will try to implement the basic elements.

![landing](https://github.com/user-attachments/assets/217fd968-ee68-4d92-b882-4fd3c26d7a17)
![protein visualization](https://github.com/user-attachments/assets/3bad6273-98a0-4f51-843b-7b93d0026500)
![info](https://github.com/user-attachments/assets/817a9c01-a503-4788-9064-8ada06530183)

**Comment by amorehead** [comment]
I love the initiative here! Thanks for kicking this off

**Comment by Poko18** [comment]
If it helps, here is already a similar implementation in [gradio](https://huggingface.co/spaces/simonduerr/RFAA) for RFAA.

**Comment by lucidrains** [comment]
@vandrw hey Andrei! i think we should just try to stretch gradio to as much as it can offer

realistically it is just a multi-input form that calls the model and renders the structure, with some other heatmaps on the side for confidence. it shouldn't be too difficult imo, but i actually was an app dev for over a decade prior to the career switch into AI.

**Comment by lucidrains** [comment]
@vandrw hey Andrei! i think we should just try to use gradio as much as possible

i'm sure there are limitations however, and we can always go for a more complete UI sans gradio at a later date, and only if we figure out some missing UI component has a big enough value add

for starters, i think just having a multi-input form for the various bio sequences + rendering the pdb and associated confidence heatmaps is good enough?

**Comment by vandrw** [comment]
@lucidrains Hi! Thanks for looking into this! I'm not worried about the graphical elements, but I was merely wondering if there are certain things we don't plan to support. I am still quite inexperienced with the AlphaFold model input and output that you have coded, and when I wrote this issue, I didn't look too closely at the `Alphafold3Input` class. However, after having a closer look, it seems I answered my own question 😄

I will start working on this once I manage to debug the template unit vector code!

**Comment by lucidrains** [comment]
@vandrw yup exactly! `Alphafold3Input` should be intuitive enough, if i designed it well

we will also need a few tests to ensure that the same `PDBInput` and `Alphafold3Input` converges to the same `AtomInput` representation. Alex and I kind of just worked off different input pipelines to parallelize work

**Comment by lucidrains** [comment]
@vandrw would say, if you haven't developed an app before, just focus on single string input protein sequence -> pdb -> rendering with 3dmol + gradio or whatever lib is popular these days. then we can slowly work our way up to a more complex form for multichain + nucleic acids + ions

**Comment by vandrw** [comment]
> re: plan to support - i haven't actually played around with their UI yet, so you'll have to let me know which things you are referring to

@lucidrains I was jumping the gun a bit here, but for example, the AlphaFold Server supports glycosylated proteins. According to their FAQ:
> The Server supports the following glycan residues to be attached to a protein residue
>  - N (Asparagine): BGC, BMA, GLC, MAN, NAG
>  - T (Threonine): BGC, BMA, FUC, GLC, MAN, NAG
>  - S (Serine): BGC, BMA, FUC, GLC, MAN, NAG
>
> The Server assumes that glycosidic bonds are formed between atoms at positions that have the highest frequency of occurrence in similar bonds from the PDB - this might lead to different bond positions in the modeled structure than expected. Specifying exact atoms for the glycosidic bond is not currently supported.

I was wondering if we need to perform any additional checks for such inputs. However, as I mentioned, I'm probably a bit ahead of myself here. This can be addressed once we have an initial version :)

Regarding my experience with app development, I've done some work in this area before. I can definitely use custom JS if the basic gradio capabilities are lacking. I will see how far I go with this! Thank you very much for your suggestions 😄

**Comment by lucidrains** [comment]
@vandrw let's definitely save modified biomolecules for last, as there is a lot of complexity there (which is mainly reflective of biology in general)

**Comment by lucidrains** [comment]
@vandrw put down some basic scaffold [here](https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/app.py)

eventually end researcher will just do `alphafold3_pytorch_app --checkpoint ./path/to/alphafold3.pt`

**Comment by Hanziwww** [comment]
Hi, I developved an Alphafold3 GUI. Project url: https://github.com/Hanziwww/AlphaFold3-GUI.
![image](https://github.com/user-attachments/assets/07b3bb47-e97a-41d0-a283-3560eaa6217f)

**Comment by lucidrains** [comment]
@Hanziwww very cool! keep up the great work!

**Comment by Hanziwww** [comment]
@lucidrains Thanks. Now I released a new version with visualization.

### 1. **Visualization Module**  
![image](https://github.com/user-attachments/assets/ae9e8b0c-8219-431f-8c1d-73f229c53039)

### 2. **Visualization GUI Interface** 
![image](https://github.com/user-attachments/assets/d91c476e-f11a-4002-a842-9ae014fce88d)

### 3. **Online Usability**    
- To make this even more accessible, I have deployed an online app where users can upload their AlphaFold3 result files and visualize them in real time.     
- This app is fully functional and available at the following link: [**Usable visualization site**](https://af3vis.streamlit.app/) **(fully usable)**
